module/state_synchronization/indexer/indexer.go

```diff
@@ -91,11 +95,29 @@ func NewIndexer(
 
 	r.jobConsumer = jobConsumer
 
+	// to separate processing execution data and updating the last indexed block
+	r.jobConsumer.SetPostNotifier(func(module.JobID) {
+		r.onBlockHeaderReceived()
+	})
+
 	r.Component = r.jobConsumer
 
 	return r, nil
 }
 
+// onBlockHeaderReceived notifies BlockHeaderDistributor that new block is indexed.
+//
+// No errors are expected during normal operations.
+func (i *Indexer) onBlockHeaderReceived() {
+	lastIndexedHeight := i.jobConsumer.LastProcessedIndex()
+	header, err := i.indexer.headers.ByHeight(lastIndexedHeight)
+	if err != nil {
+		// if the execution data is available, the block must be locally finalized
+		i.log.Error().Err(err).Msgf("could not get header for height %d:", lastIndexedHeight)
+	}
+	i.OnBlockHeaderReceived(header)
```

Comment:
I think we should handle the logic in this file a bit differently. There are 2 main issues I see now:
1. It's possible for a height to be missed here, which would result in consumers missing one or more notification. The consumer/distributor pattern should guarantee all messages are delivered.
2. We can't continue if the header was missing. A call to `OnBlockHeaderReceived` should be guaranteed to include the block header for the last index block. Otherwise, all consumers would need to add logic to handle both cases which increases the likelihood for bugs.

I'd suggest you refactor the indexer to use a `ComponentManager` instead of inheriting the `component.Component` directly from `jobConsumer`. Then it could have 2 worker:
* One that handles the lifecycle of `jobConsumer`. [here](https://github.com/onflow/flow-go/blob/master/network/p2p/node/gossipSubAdapter.go#L75-L91) is an example of how that could be done
* The other receives notifications and processes all available heights each time it receives a signal. An `engine.Notifier` is a good module to use for the signalling.

The notification loop could then get the `LastProcessedIndex()`, and call the distributor for each height since its last invocation.